src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,503 @@
+            logger.error('n_theta must be even')
+        n_theta_half = self.n_theta // 2
+
+        # The first two dimension of pf is of size n_rad x n_theta. We will convert pf
+        # into an array of size (n_rad-1) x n_theta/2, that is, take half of ray
+        # through the origin except the DC part, and also take the angles only up to PI.
+        # This is due to the fact that the original images are real, and thus each ray
+        # is conjugate symmetric. We therefore gain nothing by taking longer correlations
+        # (of length 2*n_rad-1 instead of n_rad-1). In the Matlab version, pf is convert to
+        # the size of (2*n_rad-1) x n_theta/2 but most of calculations of build_clmatrix
+        # and estimate_shifts below only use the size of (n_rad-1) x n_theta/2. In the
+        # Python version we will use the size of (n_rad-1) x n_theta/2 directly and make
+        # sure every part is using it. By taking shorter correlation we can speed the
+        # computation by a factor of two.
+        self.pf = np.flip(self.pf[1:, n_theta_half:], 0)
+
+    def estimate_rotations(self):
         """
-        pass
+        Estimate orientation matrices for all 2D images
+
+        Subclasses should implement this function.
+        """
+        raise NotImplementedError('subclasses should implement this')
+
+    def save_rotations(self):
+        """
+        Save the estimated rotation matrices to new ImageSource object
+        """
+        # Generate a new ImageSource object and save estimated rotation matrices
+        # It will be simpler if it is allowed to modify the original one
+        src = ArrayImageSource(self.src.images(start=0, num=np.inf))
+        src.rots = self.rotations
+        return src
 
     def build_clmatrix(self):
         """
-        build the common lines matrix
+        Build common-lines matrix from Fourier stack of 2D images
+        """
+
+        n_img = self.n_img
+        n_check = self.n_check
+
+        if self.n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta_half = self.n_theta // 2
+
+        # need to do a copy to prevent modifying self.pf for other functions
+        pf = self.pf.copy()
+
+        # Allocate local variables for return
+        # clmatrix represents the common lines matrix.
+        # Namely, clmatrix[i,j] contains the index in image i of
+        # the common line with image j. Note the common line index
+        # starts from 0 instead of 1 as Matlab version. -1 means
+        # there is no common line such as clmatrix[i,i].
+        clmatrix = -np.ones((n_img, n_img))
+        # cl_dist[i, j] measures how ''common'' is the between
+        # image i and j. Small value means high-similarity.
+        # we will use cl_dist[i, j] = -1 (including i=j) to
+        # represent that there is no need to check common line
+        # between i and j. Since stack is symmetric,
+        # only above the diagonal entries are necessary.
```

Comment:
I changed this variable to a local one in a preovious commit and it should not be exposed to the users. I think it is fine since there are some gains by not resetting the values when the matrix is large.

Reply:
It would have taken less characters to fix the problem.